### PR TITLE
WIP documentation updates

### PR DESCRIPTION
**Pull request by guybedford**

### CODE REVIEW

**Comment by probins** [comment]
On first glance, this looks very good, Guy. I have some comments, but I'm currently away with only a phone for Internet and will look in more detail when I'm home in a few days time.

File: docs/architecture.md [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+
+## What is the Module Loader?
+
+The module loader is a [draft specification](http://github.com/whatwg/loader) providing an API for loading modules into the page dynamically 
+after the initial page load (avoiding having to manually inject a `<script type="module">` tag into the page).
+
+But more importantly, the module loader provides an API to hook into the registry, resolution and loading pipeline of modules.
+
+The module loader is based on the principles of the [Extensible Web Manifesto](https://extensiblewebmanifesto.org/).
+By opening up the module loading primitives to allow developers deeper control over the module loading process, provides developers with
```
Comment (probins) [inline docs/architecture.md]: ... it provides ...

File: docs/architecture.md [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+The other thing to be careful of is that the polyfill itself does not interfere with the native loader when it lands in browsers.
+
+While SystemJS defines the `System` global it is recommended to use `SystemJS.import` for API calls as we can guarantee that this global name
+will not clash and conflict with the native loader when it is implemented on the `System` global in future.
+
+## SystemJS Resolution Algorithm
+
+_Everything described up until this point relates to the native module loader specification. Now we can explain the design decisions in SystemJS itself._
+
+The core of SystemJS its [resolution algorithm](resolution-algorithm.md). Previously we noted that the module specifiers as resolved like URLs,
```
Comment (probins) [inline docs/architecture.md]: ... is its ...

File: docs/architecture.md [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+(`./x`, `/x`, `//x` and `https://x` can all be considered to use URL syntax), as special _plain names_.
+
+These plain names (`import 'x'`) are reserved in the resolution to allow a custom resolution to be defined.
+
+The SystemJS resolution algorithm is inspired by the work of previous AMD loaders, and in particular RequireJS, to provide an extension of URL
+normalization that can cater to a wide variety of module resolution use cases.
+
+Unlike running traditional module resolution on a filesystem, module resolution in the browser is constrained by latency. The standard technique
+for looking up a custom name like `import _ from 'lodash'` would be to run through a series of local folder paths which may contain the module, or to
+run through a hierarchy of folders. In the browser this would result in many requests back and fourth which would not be good for user experience due to the delay.
```
Comment (probins) [inline docs/architecture.md]: ... forth ... :-)

File: docs/architecture.md [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+how to load a given module once it has been resolved to a URL.
+
+The goals here in designing SystemJS were:
+
+1. As far as possible SystemJS should be able to automatically determine the correct configuration to load a given module,
+   and this configuration should be fully exposed to the user for manually specifying how to load the module.
+2. It should be possible to load any existing module through SystemJS without having to modify the original source code.
+3. Any custom resource should be possible to load through SystemJS via a plugin system.
+4. SystemJS should work as a performant production loader, supporting configuration of
+   the optimimum network profile and loading code from bundles.
```
Comment (probins) [inline docs/architecture.md]: optimum

File: docs/architecture.md [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+5. SystemJS should enable modular package configuration o avoid the complexity that comes
```
Comment (probins) [inline docs/architecture.md]: ... to avoid ...

File: docs/architecture.md [hunk 6]
```diff
@@ -0,0 +1,448 @@
...
+
+> In future [wasm](https://en.wikipedia.org/wiki/WebAssembly) support may be provided by this same module format configuration option.
+
+#### Format Detection
+
+If every module loaded by SystemJS required the user to first indicate which of these formats it was, that would add significant
+friction to the development process so the first problem is to create a simple detection approach that is both performant and useful.
+
+Module format detection is not possible to be perfectly accurate, but SystemJS applies a detection algorithm that works as a 
+good heuristic and reliably matches the user intent in the majority of cases. This exact process is is based on regular expression detections
```
Comment (probins) [inline docs/architecture.md]: ... is ...

File: docs/architecture.md [hunk 7]
```diff
@@ -0,0 +1,448 @@
...
+This package configuration for my app can then be stored in its own JSON file at `app/systemjs.json`, and we use the `packageConfigPaths` configuration
+option to load this package configuration on-demand:
+
+```javascript
+SystemJS.config({
+  packageConfigPaths: ['app/systemjs.json']
+});
+```
+
+With the above an import to `SystemJS.import('app')` or `SystemJS.import('app/template.hbs') will detect this is a request into the package,
```
Comment (probins) [inline docs/architecture.md]: you're missing a backtick here

File: docs/module-formats.md [hunk 8]
```diff
@@ -72,7 +73,8 @@ System.import('./local-module', __moduleName);
 
 In due course this will be entirely replaced by the contextual loader once this has been specified.
 
-_ES6 is loaded via XHR making it non-[CSP](http://www.html5rocks.com/en/tutorials/security/content-security-policy/) compatible. ES6 should always be built for production to avoid transpiler costs, making this a development-only feature._
+_ES modules are loaded via XHR making it non-[CSP](http://www.html5rocks.com/en/tutorials/security/content-security-policy/) compatible.
+ES modules should always be built for production to avoid transpiler costs, making this a development-only feature._
```
Comment (probins) [inline docs/module-formats.md]: I would add 'Until `import`/`export' are natively supported' ...

**Comment by probins** [comment]
ok, some minor textual things inline. Some more general comments:

* on using modules vs scripts, I would stress the different process in handling dependencies. If x depends on y depends on z, with scripts you have to load z then y then x and make sure they are run in that sequence. Modular loading is much simpler: you just load x, and y and z are loaded automatically by the loader when it finds the import statements.
* I really don't like having jQuery in examples; jQuery is a package/script, not a module. Ok, SystemJS can load it as if it were a module, but to me a major advantage of modular loading is enabling people to only load the bits of code that are actually needed, saving people's time and bandwidth. (All those sites out there loading jQuery and then only using a couple of functions.)
* in the production workflows section, I would add a note that http/2 multiplexing removes much of the need for bundling
* I notice you don't mention the `site packages` part of the spec; perhaps mention that some of the config stuff may at some point be implemented natively?

Otherwise, I think this looks good.

**Comment by probins** [comment]
> The simplest way to load a module in the browser is through the soon-to-be-implemented module tag

not sure if you want to add these links somewhere, but if you've not seen https://groups.google.com/a/chromium.org/forum/#!msg/blink-dev/uba6pMr-jec/tXdg6YYPBAAJ it's a good summary of the situation with Chrome/Blink/V8; it also gives a link to Mozilla's work.

I would guess that these will be released behind flags for quite some time whilst everything is tested, so I don't think we should hold our breath on it being available for production any time soon :-(

File: docs/module-formats.md [hunk 9]
```diff
@@ -27,17 +27,18 @@ This module format detection is never completely accurate, but caters well for t
 
 The module format detection happens in the following order:
 * _System.register / System.registerDynamic_
-  If the source code starts with a number of comments, followed by `System.register` or `System.registerDynamic` as the first line of code.
+  If the source code starts with any number of comments, followed by `System.register` or `System.registerDynamic` as the first line of code.
 * _ES modules_
   The source is only detected as an ES module if it contains explicit module syntax - valid `import` or `export` statements.
+  This detection does not do comment removal, so commons in the code containing valid ES module syntax will trigger this detection.
```
Comment (probins) [inline docs/module-formats.md]: I think you probably mean 'comments' here rather than 'commons'

**Comment by alexisvincent** [comment]
@guybedford  Can we get this merged?

**Comment by guybedford** [comment]
@alexisvincent yes as soon as I can steal the time to complete it! Unfortunately jspm 0.17 docs for jspm release are taking priority right now (https://github.com/jspm/project/blob/master/ROADMAP.md).

**Comment by alexisvincent** [comment]
@guybedford I'm really confused as to why this was simply closed. When a docs boost would help a lot with SystemJS?

**Comment by guybedford** [comment]
@alexisvincent I brought all the changes through manually, just except for the resolution and architecture pages. A simplified resolution summary is now available at https://github.com/systemjs/systemjs/blob/master/docs/production-build.md, and the architecture page is somewhat out of date already, although it could still be updated into the docs if anyone is interested in working on this.

**Comment by alexisvincent** [comment]
(y) Makes sense. I checked to see if the resolution-algorithm.md was in the new docs (obviously had moved) and didn't see it. My bad.

**Pull request closed without merging**